javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -34,6 +39,46 @@ private PriorityQueue(SerializableComparator<T> comparator, List<Node<Ranked<T>>
     }
 
     /**
+     * Enqueues a new element.
+     *
+     * @param element The new element
+     * @return a new {@code PriorityQueue} instance, containing the new element
+     */
+    @Override
+    public PriorityQueue<T> enqueue(T element) {
+        return with(insertForest(comparator, forest, element));
+    }
+
+    /**
+     * Returns the first element of a non-empty {@link PriorityQueue}.
+     *
+     * @return The first element of this {@link PriorityQueue}.
+     * @throws NoSuchElementException if this is empty
+     */
+    @Override
+    public T head() {
+        return findMin(comparator, forest);
+    }
+
+    /**
+     * Drops the first element of a non-empty {@link PriorityQueue}.
+     *
+     * @return A new instance of PriorityQueue containing all elements except the first.
+     * @throws UnsupportedOperationException if this is empty
+     */
+    @Override
+    public PriorityQueue<T> tail() {
+        final Tuple2<List<T>, List<Node<Ranked<T>>>> tuple = deleteMinForest(comparator, forest);
+        return with(tuple._2).enqueueAll(tuple._1);
```

Comment:
internally (via the helper class) it's stateless and closer to the original functional implementation